File: tic_tac_toe/php/oo/tabuleiro.php
<?
  class Tabuleiro {
    private $tabuleiro = array(
      '     A   B   C ',
      '  1    |   |   ',
      '    ---+---+---',
      '  2    |   |   ',
      '    ---+---+---',
      '  3    |   |   '
    );
    private $jogador = 'X';
    private $matriz = array( array(false, false, false), array(false, false, false), array(false, false, false));

    public function imprimir() {
      foreach($this->tabuleiro as $t) {
        print "$t\n";
      }
      print "\nJogador $this->jogador, sua escolha? ";
    }

    public function preencher($x, $y) {
      if($this->matriz[$x][$y]) return false;
      $this->matriz[$x][$y] = $this->jogador;
      $y = $y * 2 + 1;
      $x = $x * 4 + 5;

      $this->tabuleiro[$y][$x] = $this->jogador;
      $this->jogador = $this->jogador == 'X' ? "O": "X";
      return true;
    }

    public function jogo_acabou() {
      return $this->mensagem_final();
    }

    public function mensagem_final() {
      if(!$this->m_mensagem_final) {
        $h = $this->horizontais_ou_verticais_iguais();
        if(!$h) $h = $this->diagonais_iguais();
        if($h) {
          $this->m_mensagem_final = "$h Venceu!";
        } elseif($this->todos_preenchidos()) {
          $this->m_mensagem_final = "Deu velha!";
        }
      }
      return $this->m_mensagem_final;
    }

    private function horizontais_ou_verticais_iguais() {
      foreach($this->matriz as $m) {
        $elemento = $this->iguais($m);
        if($elemento) return $elemento;
      }
      $matriz = $this->transpose($this->matriz);
      foreach($matriz as $m) {
        $elemento = $this->iguais($m);
        if($elemento) return $elemento;
      }
      return false; 
    }

    private function iguais($matriz) {
      $elemento = $matriz[0];
      for($i = 1; $i < count($matriz); $i++) {
        if($elemento != $matriz[$i]) return false;
      }
      return $elemento;
    }

    private function diagonais_iguais() {
      $matriz = $this->matriz;
      if($matriz[0][0] == $matriz[1][1] && $matriz[1][1] == $matriz[2][2]) return $matriz[0][0];
      if($matriz[0][2] == $matriz[1][1] && $matriz[1][1] == $matriz[2][0]) return $matriz[0][2];
    }

    private function transpose($original) {
      for($x = 0; $x < count($original); $x++) {
        for($y = 0; $y < count($original[$x]); $y++) {
          $transposta[$y][$x] = $original[$x][$y];
        }
      }
      return $transposta;
    }

    private function todos_preenchidos() {
      foreach($this->matriz as $m1) {
        foreach($m1 as $elemento) {
          if(!$elemento) return false;
        }
      }
      return true;
    }
  }
?>

File: tic_tac_toe/lua/oo/tabuleiro.lua
local mensagem_final

Tabuleiro = {
    '     A   B   C ',
    '  1    |   |   ',
    '    ---+---+---',
    '  2    |   |   ',
    '    ---+---+---',
    '  3    |   |   ',
    jogador = 'X',
    matriz = { { false, false, false}, { false, false, false}, { false, false, false } }
}

function Tabuleiro:imprimir()
    for i, v in ipairs(self) do print(v) end
    io.write("\nJogador " .. self.jogador .. ", sua jogada? ")
end

function Tabuleiro:preencher(x, y)
    if self.matriz[x][y] then return false end
    self.matriz[x][y] = self.jogador
    y = y * 2
    x = x * 4
    self[y] = self[y]:sub(1, x+1) .. self.jogador .. self[y]:sub(x+3)
    if self.jogador == "X" then 
        self.jogador = "O"
    else 
        self.jogador = "X"
    end
    return true
end

function Tabuleiro:jogo_acabou()
    return self:mensagem_final()
end

function Tabuleiro:mensagem_final()
    if not mensagem_final then
        h = self:horizontais_ou_verticais_iguais()
        h = h or self:diagonais_iguais()
        if h then
            mensagem_final = h .. " Venceu!"
        elseif self:todos_preenchidos() then
            mensagem_final = "Deu velha!"
        end
    end
    return mensagem_final
end

local iguais = function(matriz)
    elemento = matriz[1]
    for i = 2, 3, 1 do
        if elemento ~= matriz[i] then return nil end
    end
    return elemento
end

local transpose = function(matriz)
    transposta = { {}, {}, {} }
    for x = 1, #matriz, 1 do
        for y = 1, #matriz[x], 1 do
            transposta[y][x] = matriz[x][y]
        end
    end
    return transposta
end

function Tabuleiro:horizontais_ou_verticais_iguais()
    for k, m in ipairs(self.matriz) do
        local elemento = iguais(m)
        if elemento then return elemento end
    end

    matriz = transpose(self.matriz)
    for k, m in ipairs(matriz) do
        local elemento = iguais(m)
        if elemento then return elemento end
    end
    return nil
end

function Tabuleiro:diagonais_iguais()
    matriz = self.matriz
    if(matriz[1][1] == matriz[2][2] and matriz[2][2] == matriz[3][3]) then return matriz[1][1] end
    if(matriz[1][3] == matriz[2][2] and matriz[2][2] == matriz[3][1]) then return matriz[1][3] end
    return nil;
end

function Tabuleiro:todos_preenchidos()
    for k, m1 in ipairs(self.matriz) do
        for k2, e in ipairs(m1) do
            if(not e) then return false end
        end
    end
    return true
end
